lib/auth/permissions.go

```diff
@@ -139,18 +139,25 @@ func (a *authorizer) authorizeRemoteUser(u RemoteUser) (*Context, error) {
 	if len(roleNames) == 0 {
 		return nil, trace.AccessDenied("no roles mapped for remote user %q from cluster %q with remote roles %v", u.Username, u.ClusterName, u.RemoteRoles)
 	}
-	// Set "logins" trait and "kubernetes_groups" for the remote user. This allows Teleport to work by
-	// passing exact logins, kubernetes groups and users to the remote cluster. Note that claims (OIDC/SAML)
-	// are not passed, but rather the exact logins, this is done to prevent
-	// leaking too much of identity to the remote cluster, and instead of focus
-	// on main cluster's interpretation of this identity
+	// Set internal traits for the remote user. This allows Teleport to work by
+	// passing exact logins, kubernetes users/groups and database users/names
+	// to the remote cluster.
 	traits := map[string][]string{
 		teleport.TraitLogins:     u.Principals,
 		teleport.TraitKubeGroups: u.KubernetesGroups,
 		teleport.TraitKubeUsers:  u.KubernetesUsers,
 		teleport.TraitDBNames:    u.DatabaseNames,
 		teleport.TraitDBUsers:    u.DatabaseUsers,
 	}
+	// Prior to Teleport 6.2 no user traits were passed to remote clusters
+	// except for those specified above. To preserve backwards compatible
+	// behavior, filter out any "internal" traits when applying traits from
+	// user identity.
+	for k, v := range u.Identity.Traits {
+		if _, ok := traits[k]; !ok {
+			traits[k] = v
+		}
+	}
```

Comment:
Two questions:

1. I'm not seeing the filtering out of internal traits mentioned in the comment?
2. When can the second condition be met? Looks like you're just copying the map right? Shouldn't get duplicate keys ever?

Reply:
@russjones The filtering is the `if _, ok := traits[k]; !ok {` check - we're basically filtering out any traits which are already present in the `traits` above which contains only internal ones. In my original implementation I had an explicit function IsInternalTrait that would check if the trait is one of the "logins", "kube groups/users" or "db names/users", but then Andrew pointed out that I can check if the key is already in that `traits` map, which I did.

> When can the second condition be met?

Not sure which condition you mean? We're copying traits from the certificate essentially into the `traits` map and skipping internal ones to make sure to preserve backwards compatible behavior.